core/src/main/java/org/mule/runtime/core/internal/construct/AbstractPipeline.java

```diff
@@ -267,18 +271,25 @@ protected ReactiveProcessor processFlowFunction() {
 
       long startTime = currentTimeMillis();
 
-      // Fire COMPLETE notification on async response
-      Mono.from(((BaseEventContext) event.getContext()).getBeforeResponsePublisher())
-          .doOnSuccess(result -> fireCompleteNotification(result, null))
-          .doOnError(MessagingException.class, messagingException -> fireCompleteNotification(null, messagingException))
-          .doOnError(throwable -> !(throwable instanceof MessagingException),
-                     throwable -> fireCompleteNotification(null, new MessagingException(event, throwable,
-                                                                                        AbstractPipeline.this))
-
-          )
-          .doOnSuccessOrError((result, throwable) -> ((BaseEventContext) event.getContext()).getProcessingTime()
-              .ifPresent(time -> time.addFlowExecutionBranchTime(startTime)))
-          .subscribe(requestUnbounded());
+      BaseEventContext baseEventContext = ((BaseEventContext) event.getContext());
+      Reference<Pair<CoreEvent, Throwable>> result = new Reference<>();
+      baseEventContext.onResponse((response, throwable) -> {
+        result.set(new Pair<>(response, throwable));
+      });
+
+      baseEventContext.onComplete((a, b) -> {
```

Comment:
receive the event and throwable here instead of using the previous block to get it.